### ISSUE TITLE
Semantic search + Exact matches in ElasticSearch

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#610]
## Context
- In #609 we decided to use ElasticSearch (ES) to store our literature texts.
- ES supports various type of queries for various types of fields: `text`, `dense_vector`, ...
- Now, we would like to find out how support use cases where we want to perform complex queries that include various filters.

## Actions
- [ ] Understand how to perform complex queries, e.g. a query that contains the following filters (all filters must apply, i.e. a logical `AND` is assumed here):
   - paragraph containing the exact match `"synapses" as part of the title;
   - paragraph text semantically similar to `activity-dependent modulation of synaptic connections`;
   - date of publication more recent than `1990-01-01`.

**Comment by drsantos89** [comment]
- the match query is the default way to perform text queries on ES:
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html
- the multi-match query allows the search of multiple fields in one query:
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html
- boolean queries allows the use of a must parameter (equivalent to AND), a must_not parameter (equivalent to NOT), and a should parameter (equivalent to OR)
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html
- The simple_query_string query is a version of the query_string query that is more suitable for use in a single search box that is exposed to users because it replaces the use of AND/OR/NOT with +/|/-, respectively. It discards invalid parts of a query instead of throwing an exception if a user makes a mistake. It provides a means of executing multi_match queries, bool queries, boosting, fuzzy matching, wildcards, regexp, and range queries in a concise shorthand syntax.
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html#simple-query-string-syntax
- range queries allows the use of gt (greater than) get (greater than or equal to) lt (less than) and lte (less than or equal to). Useful for the date queries.
https://www.elastic.co/guide/en/elasticsearch/guide/current/_ranges.html

More information regarding text queries can be found at:
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html
https://dzone.com/articles/23-useful-elasticsearch-example-queries

**Comment by drsantos89** [comment]
The KNN search API can be used for semantic search:
https://www.elastic.co/guide/en/elasticsearch/reference/8.3/knn-search-api.html

The kNN search API supports restricting the search using a filter. The search will return the top k documents that also match the filter query.
https://www.elastic.co/guide/en/elasticsearch/reference/8.3/knn-search-api.html#knn-search-api-filter-example

**Comment by FrancescoCasalegno** [comment]
## Planning 2022-08-16

In view of the end-of-month Demo, prepare a couple of slides showing the following.
- A diagram of how queries for ES work (semantic search vs. classic queries by field/fulltext match) and how ML fits into this (computing vector emebeddings).
- A concrete example based on Neuroscience. E.g.: retrieve top 10 results (semantic search) about `Neuronal glucose metabolism`
- Show results for same query as previous one, but `year` of publication more recent than `2019`.
- Show results for same query as previous one, but `year` of publication more recent than `2019` and text not containing the exact match `"ATP"`.

**Comment by drsantos89** [comment]
## - Text search "Neuronal glucose metabolism"
Good results were retrieved but 2nd and 3rd hits are cancer-related.
```
query_body = {"match": {
    "abstract": "Neuronal glucose metabolism"
    },
}
resp = client.search(
    index="arxiv_dense",
    query=query_body,
    size=3,
)
print("Got %d Hits:" % resp['hits']['total']['value'])
for hit in resp['hits']['hits']:
    print(hit["_id"], hit["_score"], hit["_source"]['abstract'])
```

![Screenshot 2022-08-30 at 14 37 30](https://user-images.githubusercontent.com/9551412/187438904-791fddf2-17a4-4858-8d5c-7278157d71bd.png)

## - Semantic search "Neuronal glucose metabolism"
The 1st hit is the same. The 2nd and 3rd are now more related to the query. In the 3rd hit, none of the query words is present.
```
emb = model.encode('Neuronal glucose metabolism')
emb /= np.linalg.norm(emb)
query_body = {
    "field": "my_vector",
    "query_vector": emb,
    "k": 3,
    "num_candidates": 1000
}
resp = client.options(request_timeout=60).knn_search(
    index="arxiv_dense",
    knn=query_body,
    _source = {"includes": ["abstract", "date"]},
)
print("Got %d Hits:" % resp['hits']['total']['value'])
for hit in resp['hits']['hits']:
    print(hit["_id"], hit["_score"], hit["_source"]["date"], hit["_source"]["abstract"])
```
![Screenshot 2022-08-30 at 14 39 22](https://user-images.githubusercontent.com/9551412/187438937-f2d05501-2041-438b-832b-a6853576c853.png)

## - Semantic search "Neuronal glucose metabolism"   +   exact match "ATP"
Results were successfully filtered to include "ATP". The 3rd hit of the previous query is now the 1st.

```
emb = model.encode('Neuronal glucose metabolism')
emb /= np.linalg.norm(emb)
query_body = {
    "field": "my_vector",
    "query_vector": emb,
    "k": 3,
    "num_candidates": 1000
}
query_filter = [
    {"query_string": {
        "query": "\"ATP\"",
        "fields": ["abstract"],
        }
    }
]
resp = client.options(request_timeout=60).knn_search(
    index="arxiv_dense",
    knn=query_body,
    filter=query_filter,
    _source = {"includes": ["abstract", "date"]},
)
print("Got %d Hits:" % resp['hits']['total']['value'])
for hit in resp['hits']['hits']:
    print(hit["_id"], hit["_score"], hit["_source"]['date'], hit["_source"]['abstract'])
```
![Screenshot 2022-08-30 at 14 39 43](https://user-images.githubusercontent.com/9551412/187438959-d58307b8-ec7f-452f-ab57-460d7b358d4a.png)

## - Semantic search "Neuronal glucose metabolism"   +.  exact match "ATP"   +.  2000 <= date <= 2010
Results were successfully filtered to include "ATP" and date between 2000 and 2010. The abstract from 2018 was removed.
```
emb = model.encode('Neuronal glucose metabolism')
emb /= np.linalg.norm(emb)
query_body = {
    "field": "my_vector",
    "query_vector": emb,
    "k": 3,
    "num_candidates": 1000
}
query_filter = [
    {"query_string": {
        "query": "\"ATP\"",
        "fields": ["abstract"],
        }
    },
    {"range": {
        "date": {
            "gte": "2000",
            "lte": "2010",
        }
    }}
]
resp = client.options(request_timeout=60).knn_search(
    index="arxiv_dense",
    knn=query_body,
    filter=query_filter,
    _source = {"includes": ["abstract", "date"]},
)
print("Got %d Hits:" % resp['hits']['total']['value'])
for hit in resp['hits']['hits']:
    print(hit["_id"], hit["_score"], hit["_source"]['date'], hit["_source"]['abstract'])
```

![Screenshot 2022-08-30 at 14 39 58](https://user-images.githubusercontent.com/9551412/187438992-078a8d1e-f1a2-4fb3-8c18-86058d9c9418.png)

**Comment by FrancescoCasalegno** [comment]
Looking good to me, thanks @drsantos89 !

I would say that our use cases will mainly be "semantic search" with some sort of extra filtering (i.e. date range and/or exact match), while "text search" (despite being one of the flagship built-in features of Elasticsearch) will not really used since we will use "semantic search" instead.
What do you think about it?

**Comment by drsantos89** [comment]
Yes, I agree the "semantic search" is more powerful and will likely be the default for most cases. ES text search features might prove useful for query filtering. In this example, I have used the "query_string" to perform an exact match, but using "match", one can use the built-in ES text search features (not exact match, though).
